prow/tide/search.go

```diff
@@ -0,0 +1,147 @@
+		q = fmt.Sprintf("%s %s", q, dateToken(start, end))
+		vars := map[string]interface{}{
+			"query":        githubql.String(q),
+			"searchCursor": (*githubql.String)(nil),
+		}
+		var totalCost, remaining int
+		var totalMatches int
+		var ret []PullRequest
+		for {
+			sq := searchQuery{}
+			if err := ghc.Query(ctx, &sq, vars); err != nil {
+				return nil, 0, err
+			}
+			totalCost += int(sq.RateLimit.Cost)
+			remaining = int(sq.RateLimit.Remaining)
+			totalMatches = int(sq.Search.IssueCount)
+			// If the search won't return all results, abort.
+			if totalMatches > 1000 {
+				return nil, totalMatches, nil
+			}
+			for _, n := range sq.Search.Nodes {
+				ret = append(ret, n.PullRequest)
+			}
+			if !sq.Search.PageInfo.HasNextPage {
+				break
+			}
+			vars["searchCursor"] = githubql.NewString(sq.Search.PageInfo.EndCursor)
+		}
+		log.Debugf("Search for query \"%s\" returned %d PRs and cost %d point(s). %d remaining.", q, len(ret), totalCost, remaining)
+		return ret, totalMatches, nil
+	}
+}
+
+func (q searchExecutor) search() ([]PullRequest, error) {
+	prs, _, err := q.searchRange(time.Time{}, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchSince(t time.Time) ([]PullRequest, error) {
+	prs, _, err := q.searchRange(t, time.Now())
+	return prs, err
+}
+
+func (q searchExecutor) searchRange(start, end time.Time) ([]PullRequest, int, error) {
+	prs, count, err := q(start, end)
+	if err != nil {
+		return nil, 0, err
+	}
+
+	if count <= 1000 {
+		// The search returned all the results for the query.
+		return prs, len(prs), nil
+	}
+	// The query returned too many results, we need to partition it.
+	prs, err = q.partitionSearchRange(start, end, count)
+	return prs, len(prs), err
+}
+
+func (q searchExecutor) partitionSearchRange(start, end time.Time, count int) ([]PullRequest, error) {
+	partition := partitionTime(start, end, count, 900)
+	// Search right side...
+	rPRs, rCount, err := q.searchRange(partition, end)
+	if err != nil {
+		return nil, err
+	}
+
+	// Search left side...
+	// For the left side we can deduce the count in advance.
+	lCount := count - rCount
+	// If the count is too large we can skip the initial search and go straight to
+	// partitioning to save an API token.
+	var lPRs []PullRequest
+	if lCount <= 1000 {
+		lPRs, _, err = q.searchRange(start, partition)
+	} else {
+		lPRs, err = q.partitionSearchRange(start, partition, lCount)
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	return append(lPRs, rPRs...), nil
+}
+
+func partitionTime(start, end time.Time, count, goalSize int) time.Time {
+	// TODO(cjwagner): Special case old start times. (Nobody has PRs from before 2000)
+	duration := end.Sub(start)
+	if count < goalSize*2 {
+		// Choose the midpoint.
+		return start.Add(duration / 2)
+	}
+	// Choose the point that will make the partitionTime->end range contain goalSize
+	// many results assuming a uniform distribution over time.
```

Comment:
Yep, it's highly weighted towards recent times FWIW.

k/k open PRs as of tonight (numbers are cumulative):
2 <= 2016 (two September PRs)
5 <= 2017-03 (3 more PRs)
10 <= 2017-06
22 <= 2017-09
68 <=2017-12
210 <= 2018-03
467 <= 2018-06
940 <= 2018-09

So at the midpoint of roughly 1 year ago we have 22 PRs before and 918 more after.
Similarly we have 5 and 17 when we divide again, and 188 and 708 dividing the more recent  half.

Reply:
Though we should really only need to divide ~once for the foreseeable future. k/k has had more than 1k prs before, but not a lot more, and is back under currently.

Reply:
This isn't per repo though. The query for all open PRs across all repos has 2200+ results. We shouldn't need to divide more than a few times though and this will only affect the first status sync loop as subsequent loops will only cover changes since the last loop.

Reply:
Sure, "assumes a uniform distribution" is likely an inaccurate assumption though. (the code should work fine, but this is definitely not true).

Reply:
SGTM